tools/PinVersions/PinVersionUtility.cs

```diff
@@ -40,32 +40,83 @@ public PinVersionUtility(string repositoryRoot, List<string> pinSources, Depende
+                    projectFileInfo.Name + ".pinnedversions.targets");
+
+                Directory.CreateDirectory(Path.GetDirectoryName(pinnedReferencesFile));
+
+                Console.WriteLine($"Pinning package versions for {specProject.FilePath}.");
+                var pinnedReferences = new XElement("ItemGroup");
+                foreach (var packageReference in projectPinMetadata.Packages)
+                {
+                    (var tfm, var libraryRange, var exactVersion) = packageReference;
+                    Console.WriteLine($"Pinning reference {libraryRange.Name}({libraryRange.VersionRange} to {exactVersion}.");
+                    var metadata = new List<XAttribute>
+                    {
+                        new XAttribute("Update", libraryRange.Name),
+                        new XAttribute("Version", exactVersion.ToNormalizedString()),
+                    };
+
+                    if (tfm != NuGetFramework.AnyFramework)
+                    {
+                        metadata.Add(new XAttribute("Condition", $"'$(TargetFramework)'=='{tfm.GetShortFolderName()}'"));
+                    }
+
+                    pinnedReferences.Add(new XElement("PackageReference", metadata));
+                }
+
+                foreach (var toolReference in projectPinMetadata.CLIToolReferences)
+                {
+                    (var libraryRange, var exactVersion) = toolReference;
+                    Console.WriteLine($"Pinning CLI Tool {libraryRange.Name}({libraryRange.VersionRange} to {exactVersion}.");
+                    var metadata = new List<XAttribute>
+                    {
+                        new XAttribute("Update", libraryRange.Name),
+                        new XAttribute("Version", exactVersion.ToNormalizedString()),
+                    };
+
+                    pinnedReferences.Add(new XElement("DotNetCliToolReference", metadata));
+                }
+
+                var pinnedVersionRoot = new XElement("Project", pinnedReferences);
+                File.WriteAllText(pinnedReferencesFile, pinnedVersionRoot.ToString());
+            }
+        }
+
+        private IDictionary<string, PinVersionMetadata> GetPinVersionMetadata()
+        {
             var repositoryDirectoryInfo = new DirectoryInfo(_repositoryRoot);
-            var knownProjects = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            var projects = new Dictionary<string, PinVersionMetadata>(StringComparer.OrdinalIgnoreCase);
+
             foreach (var slnFile in repositoryDirectoryInfo.EnumerateFiles("*.sln"))
             {
                 var graphSpec = _provider.GetDependencyGraphSpec(repositoryDirectoryInfo.Name, slnFile.FullName);
                 foreach (var specProject in graphSpec.Projects)
                 {
-                    if (!knownProjects.Add(specProject.FilePath) ||
-                       specProject.RestoreMetadata.ProjectStyle != ProjectStyle.PackageReference)
```

Comment:
Here's where we were skipping `DotNetCliToolReference`s previously. We now classify these references correctly and create either a `<PackageReference Update=".." Version="..." />` node or a `<DotNetCliToolReference Update=".." Version="..." />`